internal/source/source.go

```diff
@@ -0,0 +1,67 @@
+package source
+
+import (
+	"context"
+	"errors"
+	"io/fs"
+	"sync"
+
+	rukpakv1alpha1 "github.com/operator-framework/rukpak/api/v1alpha1"
+)
+
+type Source interface {
+	SourceContent(context.Context, *rukpakv1alpha1.Bundle) (*Result, error)
+}
+
+type Result struct {
+	Bundle         fs.FS
+	ResolvedSource *rukpakv1alpha1.BundleSource
+	State          State
+	Message        string
+}
+
+type State string
+
+const (
+	StatePending   State = "Pending"
+	StateUnpacking State = "Unpacking"
+	StateUnpacked  State = "Unpacked"
+)
+
+type Sourcer struct {
+	sources map[rukpakv1alpha1.BundleSourceType]Source
+	m       sync.RWMutex
+}
+
+var (
+	ErrSourceAlreadyAdded = errors.New("source type already added")
+	ErrSourceNotSupported = errors.New("source type not supported")
+)
+
+func NewSourcer() *Sourcer {
+	return &Sourcer{sources: map[rukpakv1alpha1.BundleSourceType]Source{}}
+}
+
+func (s *Sourcer) Add(typ rukpakv1alpha1.BundleSourceType, source Source) error {
+	s.m.RLock()
+	_, ok := s.sources[typ]
+	s.m.RUnlock()
+	if ok {
+		return ErrSourceAlreadyAdded
+	}
+
```

Comment:
nit: I assume that this mutex for thread safe.  If 2 threads for the same `typ` come here at the same time and both get the `ok` false, the source of one thread is overwritten by the other thread without an error.

Reply:
I updated this logic to use a write lock around the entire function. That way by the time the second thread enters the critical section, the map will include the contents of the first thread's source and that will result in the second thread getting the error.

Reply:
I've completely removed the mutex in the latest commit, which is made possible by forcing the set of sources to be provided as a map only during initialization of the object. New sources can no longer be added later.